ddtrace/contrib/unittest/patch.py

```diff
@@ -66,31 +144,128 @@ def _extract_module_name_from_test_method(item):
+
+
+def _is_valid_test_call(kwargs):
+    return not len(kwargs)
+
+
+def _is_valid_module_suite_call(func):
+    return type(func).__name__ == "method" or type(func).__name__ == "instancemethod"
+
+
+def _is_invoked_by_cli(instance):
+    return (
+        hasattr(instance, "progName")
+        or hasattr(_CIVisibility, "_datadog_entry")
+        and _CIVisibility._datadog_entry == "cli"
+    )
+
+
+def _is_invoked_by_text_test_runner():
+    return hasattr(_CIVisibility, "_datadog_entry") and _CIVisibility._datadog_entry == "TextTestRunner"
+
+
+def _generate_module_suite_path(test_module_path, test_suite_name):
+    return "{}.{}".format(test_module_path, test_suite_name)
+
+
+def _populate_suites_and_modules(test_objects, seen_suites, seen_modules):
+    if not hasattr(test_objects, "__iter__"):
+        return
+    for test_object in test_objects:
+        if not _is_test(test_object):
+            _populate_suites_and_modules(test_object, seen_suites, seen_modules)
+            continue
+        test_module_path = _extract_module_file_path(test_object)
+        test_suite_name = _extract_suite_name_from_test_method(test_object)
+        test_module_suite_path = _generate_module_suite_path(test_module_path, test_suite_name)
+        if test_module_path not in seen_modules:
+            seen_modules[test_module_path] = {
+                "module_span": None,
+                "remaining_suites": 0,
+            }
+        if test_module_suite_path not in seen_suites:
+            seen_suites[test_module_suite_path] = {
+                "suite_span": None,
+                "remaining_tests": 0,
+            }
+
+            seen_modules[test_module_path]["remaining_suites"] += 1
+
+        seen_suites[test_module_suite_path]["remaining_tests"] += 1
+
+
+def _finish_remaining_suites_and_modules(seen_suites, seen_modules):
+    for suite in seen_suites.values():
+        if suite["suite_span"] and not suite["suite_span"].finished:
+            _update_status_item(suite["suite_span"]._parent, suite["suite_span"].get_tag(test.STATUS))
+            suite["suite_span"].finish()
+
+    for module in seen_modules.values():
+        if module["module_span"] and not module["module_span"].finished:
+            _update_status_item(module["module_span"]._parent, module["module_span"].get_tag(test.STATUS))
+            module["module_span"].finish()
+    del _CIVisibility._unittest_data
+
+
+def _update_remaining_suites_and_modules(test_module_suite_path, test_module_path, test_module_span, test_suite_span):
+    _CIVisibility._unittest_data["suites"][test_module_suite_path]["remaining_tests"] -= 1
+    if _CIVisibility._unittest_data["suites"][test_module_suite_path]["remaining_tests"] == 0:
+        _CIVisibility._unittest_data["modules"][test_module_path]["remaining_suites"] -= 1
+        _finish_span(test_suite_span)
+    if _CIVisibility._unittest_data["modules"][test_module_path]["remaining_suites"] == 0:
+        _finish_span(test_module_span)
```

Comment:
This method has many nested dictionary look ups. Can this fail in unexpected ways? Should we wrap this in a try-except and log errors?

Reply:
It should not, we're checking if `test_module_path` and `test_module_suite_path` exists in the `_CIVisibility._unittest_data` dictionary [here](https://github.com/DataDog/dd-trace-py/blob/a7b14cb8a7239134b39995d15e55ba476fe93df0/ddtrace/contrib/unittest/patch.py#L381)

Reply:
Beyond the potential for failure, this code is kind of hard to read. Intermediate variables could help with that.

Reply:
I added some new variables to clear up this function, hopefully it is more readable now. Thank you!